### PR TITLE
Refactor kubernetes test

### PR DESCRIPTION
**Pull request by bricef**
Refactoring the fairly complex test script for weave & Kubernetes to be broken up into functions. 

Have rebased and broken into smaller commits to intentions clear and clear up a couple of mistakes during the original refactor. 

Semantics should remain identical.

### CODE REVIEW

File: test/README.md [hunk 1]
```diff
@@ -91,3 +91,62 @@ When a new version of Docker is released, you willneed to update the GCE test im
 To do this, change the Docker version in `run-integration-tests.sh` and push the change.
 Next build in CircleCI will detect that there is no template for this version of Docker and will first create the template before running tests.
 Subsequent builds will then simply re-use the template.
+
+## Running the tests on a cloud provider
+Running the tests against a remote provider instead of local vagrant VMs is a little more involved, and reproduces the environment used for CI. It is recommended that you use the build VM to coordinate the build phase.
+
+1. Create the build VM and log in
```
Comment (brb) [inline test/README.md]: To avoid possible confusions, maybe it's worth specifying from which dir to create the VM, i.e. either `$WEAVE_ROOT` or `$WEAVE_ROOT/test`?

File: test/README.md [hunk 2]
```diff
@@ -91,3 +91,62 @@ When a new version of Docker is released, you willneed to update the GCE test im
 To do this, change the Docker version in `run-integration-tests.sh` and push the change.
 Next build in CircleCI will detect that there is no template for this version of Docker and will first create the template before running tests.
 Subsequent builds will then simply re-use the template.
+
+## Running the tests on a cloud provider
+Running the tests against a remote provider instead of local vagrant VMs is a little more involved, and reproduces the environment used for CI. It is recommended that you use the build VM to coordinate the build phase.
```
Comment (brb) [inline test/README.md]: Consistency nit-pick: empty line before the paragraph.

File: test/README.md [hunk 3]
```diff
@@ -91,3 +91,62 @@ When a new version of Docker is released, you willneed to update the GCE test im
...
+
+1. Create the build VM and log in
+    
+    ```
+    ➤ vagrant up
```
Comment (brb) [inline test/README.md]: Nit-pick: s/➤/$/ as it is more common.

File: test/README.md [hunk 4]
```diff
@@ -91,3 +91,62 @@ When a new version of Docker is released, you willneed to update the GCE test im
...
+
+    ```
+    vagrant@vagrant:~$ export SECRET_KEY="XXXXXXXXXX"
+    ```
+
+4. Navigate to the test directory and provision the VMs
+
+    ```
+    vagrant@vagrant:~$ cd weave/test
+    vagrant@vagrant:~/weave/test$ ./run-integration-tests.sh provision
```
Comment (brb) [inline test/README.md]: What are names of the provisioned VMs? Asking as there is a GC in the wild periodically destroying VMs which names match https://github.com/weaveworks/build-tools/blob/master/scheduler/main.py#L109

File: test/840_weave_kube_3_test.sh [hunk 5]
```diff
@@ -131,30 +120,62 @@ spec:
...
+    # Destroy our test ipset, and implicitly check it is still there
+    assert_raises "docker_on $HOST1 run --rm --privileged --net=host --entrypoint=/usr/sbin/ipset weaveworks/weave-npc destroy test_840_ipset"
+
 }
 
-assert_raises 'wait_for_x check_all_pods_communicate pods'
+function cleanup {
+teardown_kubernetes_cluster
+}
+trap cleanup EXIT
```
Comment (brb) [inline test/840_weave_kube_3_test.sh]: Won't it override another `EXIT` handler https://github.com/weaveworks/weave/blob/master/test/assert.sh#L186?

File: test/840_weave_kube_3_test.sh [hunk 6]
```diff
@@ -20,59 +15,56 @@ KUBECTL="sudo kubectl --kubeconfig /etc/kubernetes/admin.conf"
...
 fi
 
-# Ensure Kubernetes uses locally built container images and inject code coverage environment variable (or do nothing depending on $COVERAGE):
-sed -e "s%imagePullPolicy: Always%imagePullPolicy: Never%" \
-    -e "s%env:%$COVERAGE_ARGS%" \
-    "$(dirname "$0")/../prog/weave-kube/weave-daemonset-k8s-1.7.yaml" | run_on "$HOST1" "$KUBECTL apply -n kube-system -f -"
+#
+# Utility functions
+#
+tear_down_kubeadm() {
```
Comment (brb) [inline test/840_weave_kube_3_test.sh]: For sake of consistency, maybe you could add `function` keyword before the function name? Applies to other functions as well.

File: test/840_weave_kube_3_test.sh [hunk 7]
```diff
@@ -131,30 +120,62 @@ spec:
...
-assert_raises 'wait_for_x check_all_pods_communicate pods'
+function cleanup {
+teardown_kubernetes_cluster
+}
+trap cleanup EXIT
 
-# Check that a pod can contact the outside world
-assert_raises "$SSH $HOST1 $KUBECTL exec $podName -- $PING 8.8.8.8"
+function main {
+    start_suite "Test weave-net deallocates IPAM on node failure";
```
Comment (brb) [inline test/840_weave_kube_3_test.sh]: The same test suite is used to test weave-kube, weave-npc, so the suite name is a bit misleading.

File: test/840_weave_kube_3_test.sh [hunk 8]
```diff
@@ -131,30 +120,62 @@ spec:
...
 
-assert_raises 'wait_for_x check_all_pods_communicate pods'
+function cleanup {
+teardown_kubernetes_cluster
+}
+trap cleanup EXIT
 
-# Check that a pod can contact the outside world
-assert_raises "$SSH $HOST1 $KUBECTL exec $podName -- $PING 8.8.8.8"
+function main {
```
Comment (brb) [inline test/840_weave_kube_3_test.sh]: Maybe it's a personal taste, but I prefer to have assertions in locations as they used to be (unless they are repeated). Having them in the functions requires a bit more context switching
when reading the suite.

Also, IMHO, `main` is a bit an overkill.

File: test/840_weave_kube_3_test.sh [hunk 9]
```diff
@@ -20,59 +15,56 @@ KUBECTL="sudo kubectl --kubeconfig /etc/kubernetes/admin.conf"
...
+
 
-sleep 5
 
 check_connections() {
     run_on $HOST1 "curl -sS http://127.0.0.1:6784/status | grep \"$SUCCESS\""
 }
 
-assert_raises 'wait_for_x check_connections "connections to establish"'
+check_ready() {
```
Comment (brb) [inline test/840_weave_kube_3_test.sh]: Not related to your PR, but because you touched the suite ;-) : maybe you could make some function names more accurate. E.g., `check_ready` does not say what supposed to be ready.

File: test/README.md (hunk 4 above)
Comment (bricef) [reply test/README.md]: @brb Provisioned VMs are by default `vagr-0`,`vagr-1`, etc... They don't seem to be picked up by any garbage collection so far. They would definitely _not_ be picked up by GC for anyone not using the Weaveworks GCP account

File: test/840_weave_kube_3_test.sh (hunk 7 above)
Comment (bricef) [reply test/840_weave_kube_3_test.sh]: 👍 You're right. That's a mistake.

File: test/840_weave_kube_3_test.sh (hunk 6 above)
Comment (bricef) [reply test/840_weave_kube_3_test.sh]: 👍

File: test/840_weave_kube_3_test.sh (hunk 5 above)
Comment (bricef) [reply test/840_weave_kube_3_test.sh]: 👍 Double checked this and it will. I've fixed this by using some bash subshells to get the behaviour I wanted.

File: test/840_weave_kube_3_test.sh [hunk 10]
```diff
@@ -20,59 +15,56 @@ KUBECTL="sudo kubectl --kubeconfig /etc/kubernetes/admin.conf"
...
+function check_all_pods_communicate {
+    assert_raises 'wait_for_x can_pods_communicate pods'
 
-# Check we can ping between the Weave bridg IPs on each host
-HOST1EXPIP=$($SSH $HOST1 "weave expose" || true)
-HOST2EXPIP=$($SSH $HOST2 "weave expose" || true)
-HOST3EXPIP=$($SSH $HOST3 "weave expose" || true)
-assert_raises "run_on $HOST1 $PING $HOST2EXPIP"
-assert_raises "run_on $HOST2 $PING $HOST1EXPIP"
-assert_raises "run_on $HOST3 $PING $HOST2EXPIP"
```
Comment (bboreham) [inline test/840_weave_kube_3_test.sh]: Removing the ping tests is rather beyond a "refactor"; if it is intentional please do it in a separate commit and explain.

File: test/840_weave_kube_3_test.sh [hunk 11]
```diff
@@ -20,59 +15,56 @@ KUBECTL="sudo kubectl --kubeconfig /etc/kubernetes/admin.conf"
...
+    # Check that a pod can contact the outside world
+    assert_raises "$SSH $HOST1 $KUBECTL exec $podName -- $PING 8.8.8.8"
 
-# Ensure we do not generate any defunct process (e.g. launch.sh) after starting weaver:
-assert "run_on $HOST1 ps aux | grep -c '[d]efunct'" "0"
-assert "run_on $HOST2 ps aux | grep -c '[d]efunct'" "0"
-assert "run_on $HOST3 ps aux | grep -c '[d]efunct'" "0"
```
Comment (bboreham) [inline test/840_weave_kube_3_test.sh]: why have these checks disappeared?

File: test/840_weave_kube_3_test.sh [hunk 12]
```diff
@@ -20,59 +15,56 @@ KUBECTL="sudo kubectl --kubeconfig /etc/kubernetes/admin.conf"
...
 }
 
-assert_raises 'wait_for_x check_connections "connections to establish"'
+check_ready() {
+    run_on $HOST1 "$KUBECTL get nodes | grep -c -w Ready | grep $NUM_HOSTS"
+}
+
+function can_pods_communicate {
+    podName=$($SSH $HOST1 "$KUBECTL get pods -l run=nettest -o go-template='{{(index .items 0).metadata.name}}'")
+    status=$($SSH $HOST1 "$KUBECTL exec $podName -- curl -s -S http://127.0.0.1:8080/status")
```
Comment (bboreham) [inline test/840_weave_kube_3_test.sh]: The semantics here have changed from checking repeatedly that one pod has passed the test to checking that a different pod each time has passed the test.
I think there is more chance of a false positive with this change.

Comment (bricef) [reply test/840_weave_kube_3_test.sh]: @bboreham Will 
```
$SSH $HOST1 "$KUBECTL get pods -l run=nettest -o go-template='{{(index .items 0).metadata.name}}'"
```
not always result in the same pod name?

File: test/840_weave_kube_3_test.sh (hunk 10 above)
Comment (bricef) [reply test/840_weave_kube_3_test.sh]: @bboreham Thank you, good catch. I didn't intend to remove those in 840, but in another test.

File: test/840_weave_kube_3_test.sh (hunk 11 above)
Comment (bricef) [reply test/840_weave_kube_3_test.sh]: @bboreham same as above, unintentional removals!

**Comment by brb** [comment]
This PR is quite out dated. Is anyone working on it?

**Comment by bricef** [comment]
@brb Yes, it's in my backlog and I'll need to re-work it to include current test changes.

**Comment by bricef** [comment]
This is superseded by #3218, I'm closing this PR.
